### ISSUE TITLE
CrossBatchMemory forward method argument names aren't the same as in BaseMetricLossFunction

### ISSUE THREAD

**Issue by thinline72** [KevinMusgrave/pytorch-metric-learning#127]
Hi,
First of all: thanks a lot for the awesome library!

I've just started trying CrossBatchMemory in my project. I activate it after some number of training epochs by just wrapping the current loss function and miner with it. But first try failed as CrossBatchMemory uses [input_indices_tuple](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/cross_batch_memory.py#L15) attribute name when other losses use [indices_tuple](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/base_metric_loss_function.py#L19).

For sure removing argument names where the loss is applied helps, but I'm wondering shouldn't it follow the same naming? 
Or maybe I'm just doing something wrong (as I wasn't able to get reasonable results with CrossBatchMemory yet).

**Comment by KevinMusgrave** [comment]
You're right, it would make more sense to use the same argument name (input_indices) as all the other loss functions. I'll add this to my to-do list.

**Comment by KevinMusgrave** [comment]
Does the accuracy keep rising after you start using CrossBatchMemory, or is something catastrophic happening, like the accuracy collapses?

**Comment by thinline72** [comment]
Ranking metrics actually drops quite a lot comparing to the same setup but without CrossBatchMemory, especially with large memory ratio. But I'm still playing with loss/miner params and number of warmup epochs, trying to get a feeling of how it works altogether.

I also noticed that loss drops a lot but seems like I was able to eliminate that by using AvgNonZeroReducer. So far I mostly experimenting with MultiSimilarity with the following params:
```
    miner = miners.MultiSimilarityMiner(epsilon=0.1)
    metric_loss = losses.MultiSimilarityLoss(
        alpha=2, beta=40, base=0.5, reducer=reducers.AvgNonZeroReducer()
    )
```

BTW, could you please help me to understand what happens [here](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/cross_batch_memory.py#L65-L70). As far as I see miner can be used inside CrossBatchMemory to mine from the memory as well as within the batch outside from CrossBatchMemory. Then results are concatenated. Am I right here?

Another question is why we concat current batch embs/labels with the memory [here](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/cross_batch_memory.py#L31-L32)? Assuming that we already just added the current batch into memory [here](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/cross_batch_memory.py#L21)

**Comment by thinline72** [comment]
I also noticed that DistanceWeightedMiner doesn't work with CrossBatchMemory although it's also a tuple miner. 
Sorry, lost the logs, but I'll try it again and create another issue then.

**Comment by KevinMusgrave** [comment]
> BTW, could you please help me to understand what happens [here](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/cross_batch_memory.py#L65-L70). As far as I see miner can be used inside CrossBatchMemory to mine from the memory as well as within the batch outside from CrossBatchMemory. Then results are concatenated. Am I right here?

Yep that's correct

> Another question is why we concat current batch embs/labels with the memory [here](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/cross_batch_memory.py#L31-L32)? Assuming that we already just added the current batch into memory [here](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/master/src/pytorch_metric_learning/losses/cross_batch_memory.py#L21)

When we add embeddings to the memory queue, they get detached from autograd. So we can't use only the memory queue to compute the loss. 

However, we could add the current batch of embeddings to the queue **after** the loss is computed. This is actually how I used to have it implemented because it was the most intuitive approach. But it [doesn't work well as seen in this issue](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/24). 

Also, the [paper](https://arxiv.org/pdf/1912.06798.pdf) explicitly says to add the current batch to the queue before computing the loss: 
![image](https://user-images.githubusercontent.com/2314891/85317828-0b3a9b00-b48d-11ea-8f9c-394ea48e4b29.png)

**Comment by thinline72** [comment]
Ah thanks a lot! Sorry, missed that detail with detaching, totally makes sense.

So if I got it right, if I provide miner both inside CrossBatch and outside, then the relationships within the current batch will be considered twice. But depending on miner they might be considered in a different way.

So far no luck with getting better results, but I'll continue experiments! Thanks!

**Comment by KevinMusgrave** [comment]
> So if I got it right, if I provide miner both inside CrossBatch and outside, then the relationships within the current batch will be considered twice. But depending on miner they might be considered in a different way.
Yes

**Comment by KevinMusgrave** [comment]
Changed ```input_indices_tuple``` to ```indices_tuple``` in v0.9.89.dev2

**Comment by thinline72** [comment]
Just an update on my experiments with CrossBatchMemory:
It didn't work for me if I train network (Embeddings + RNN) from scratch and if I use a big enough batches like 256+. But it worked very well and improved results if I fine-tune big networks like BERT with small batch sizes (8-32) and low LR like 1e-5. In that case it works even starting from the beginning of training, but I usually wait at least one epoch with LR warmup and then activate it.

**Comment by KevinMusgrave** [comment]
@thinline72 Thanks for the update! When you say batch size, do you mean the memory queue size, or the batch size of your dataloader?

**Comment by thinline72** [comment]
@KevinMusgrave I was referring to the batch size of my loader. And this seems to be similar to the conclusions in the paper. The smaller batch size the less positives/negatives we can mine within it. But cross-batch memory acts like a bigger batch size here allowing to find better positives/negatives. At least this is how I see it 🙂 

For the memory size I used values from 256-1024 depending on the batch size 8-32.
